README.md
dalek-driver-sauce
===================

> DalekJS driver bindings for [Sauce Labs](https://saucelabs.com) integration

[![Build Status](https://travis-ci.org/dalekjs/dalek-driver-sauce.png)](https://travis-ci.org/dalekjs/dalek-driver-sauce)
[![Build Status](https://drone.io/github.com/dalekjs/dalek-driver-sauce/status.png)](https://drone.io/github.com/dalekjs/dalek-driver-sauce/latest)
[![Dependency Status](https://david-dm.org/dalekjs/dalek-driver-sauce.png)](https://david-dm.org/dalekjs/dalek-driver-sauce)
[![devDependency Status](https://david-dm.org/dalekjs/dalek-driver-sauce/dev-status.png)](https://david-dm.org/dalekjs/dalek-driver-sauce#info=devDependencies)
[![NPM version](https://badge.fury.io/js/dalek-driver-sauce.png)](http://badge.fury.io/js/dalek-driver-sauce)
[![Coverage](http://dalekjs.com/package/dalek-driver-sauce/master/coverage/coverage.png)](http://dalekjs.com/package/dalek-driver-sauce/master/coverage/index.html)
[![unstable](https://rawgithub.com/hughsk/stability-badges/master/dist/unstable.svg)](http://github.com/hughsk/stability-badges)
[![Stories in Ready](https://badge.waffle.io/dalekjs/dalek-driver-sauce.png?label=ready)](https://waffle.io/dalekjs/dalek-driver-sauce)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/dalekjs/dalek-driver-sauce/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
[![Built with Grunt](https://cdn.gruntjs.com/builtwith.png)](http://gruntjs.com/)

[![NPM](https://nodei.co/npm/dalek-driver-sauce.png)](https://nodei.co/npm/dalek-driver-sauce/)
[![NPM](https://nodei.co/npm-dl/dalek-driver-sauce.png)](https://nodei.co/npm/dalek-driver-sauce/)

## Ressources

[API Docs](http://dalekjs.com/package/dalek-driver-sauce/master/api/index.html) -
[Trello](https://trello.com/b/896PxIhS/dalek-driver-sauce) -
[Code coverage](http://dalekjs.com/package/dalek-driver-sauce/master/coverage/index.html) -
[Code complexity](http://dalekjs.com/package/dalek-driver-sauce/master/complexity/index.html) -
[Contributing](https://github.com/dalekjs/dalek-driver-sauce/blob/master/CONTRIBUTING.md) -
[User Docs](http://dalekjs.com/docs/sauce.html) -
[Homepage](http://dalekjs.com) -
[Twitter](http://twitter.com/dalekjs)

## Docs

This module is a driver plugin for [DalekJS](//github.com/dalekjs/dalek).
It connects Daleks testsuite with the remote testing environment of [Sauce Labs](https://saucelabs.com).

The driver can be installed with the following command:

```
$ npm install dalek-driver-sauce --save-dev
```

You can use the driver by adding a config option to the your [Dalekfile](http://dalekjs.com/docs/config.html)

```js
"driver": ["sauce"]
```

Or you can tell Dalek that it should run your tests via sauces service via the command line:

```
$ dalek mytest.js -d sauce
```

In order to run your tests within the Sauce Labs infrastructure, you must add your sauce username & key 
to your dalek configuration. Those two parameters must be set in order to get this driver up & running.
You can specifiy them within your [Dalekfile](http://dalekjs.com/docs/config.html) like so:

```js
"driver.sauce": {
  "user": "dalekjs",
  "key": "aaaaaa-1234-567a-1abc-1br6d9f68689"
}
```

It is also possible to specify a set of other extra saucy parameters like `name` & `tags`:

```js
"driver.sauce": {
  "user": "dalekjs",
  "key": "aaaaaa-1234-567a-1abc-1br6d9f68689",
  "name": "Guineapig",
  "tags": ["dalek", "testproject"]
}
```

If you would like to have a more control over the browser/OS combinations that are available, you are able 
to configure you custom combinations:

```js
"browsers": [{
  "chrome": {
    "platform": "OS X 10.6",
    "actAs": "chrome",
    "version": 27
  },
  "chromeWin": {
    "platform": "Windows 7",
    "actAs": "chrome",
    "version": 27
  },
  "chromeLinux": {
    "platform": "Linux",
    "actAs": "chrome",
    "version": 26
  }
```

You can then call your custom browsers like so:

```
$ dalek mytest.js -d sauce -b chrome,chromeWin,chromeLinux
```

or you can define them in your Dalekfile:

```js
"browser": ["chrome", "chromeWin", "chromeLinux"]
```

A list of all available browser/OS combinations, can be found [here](https://saucelabs.com/docs/platforms).

## Help Is Just A Click Away

### #dalekjs on FreeNode.net IRC

Join the `#daleksjs` channel on [FreeNode.net](http://freenode.net) to ask questions and get help.

### [Google Group Mailing List](https://groups.google.com/forum/#!forum/dalekjs)

Get announcements for new releases, share your projects and ideas that are
using DalekJS, and join in open-ended discussion that does not fit in
to the Github issues list or StackOverflow Q&A.

**For help with syntax, specific questions on how to implement a feature
using DalekJS, and other Q&A items, use StackOverflow.**

### [StackOverflow](http://stackoverflow.com/questions/tagged/dalekjs)

Ask questions about using DalekJS in specific scenarios, with
specific features. For example, help with syntax, understanding how a feature works and
how to override that feature, browser specific problems and so on.

Questions on StackOverflow often turn in to blog posts or issues.

### [Github Issues](//github.com/dalekjs/dalek-driver-sauce/issues)

Report issues with DalekJS, submit pull requests to fix problems, or to
create summarized and documented feature requests (preferably with pull
requests that implement the feature).

**Please don't ask questions or seek help in the issues list.** There are
other, better channels for seeking assistance, like StackOverflow and the
Google Groups mailing list.

![DalekJS](https://raw.github.com/dalekjs/dalekjs.com/master/img/logo.png)

## Legal FooBar (MIT License)

Copyright (c) 2013 Sebastian Golasch

Distributed under [MIT license](https://github.com/dalekjs/dalek-driver-sauce/blob/master/LICENSE-MIT)

CHANGELOG.md
<a name="0.0.1"></a>
### 0.0.1 (2014-11-28)




test/index_TEST.js
'use strict';

var expect = require('chai').expect;
var driver = require('../index.js');

describe('dalek-driver-sauce', function() {

  it('i should think of proper testing', function(){
    expect(driver).to.be.ok;
  });

});


test/lib_browser_TEST.js
'use strict';

var expect = require('chai').expect;
var browser = require('../lib/browser.js');

describe('dalek-driver-sauce Browser', function() {

  it('i should think of proper testing', function(){
    expect(browser).to.be.ok;
  });

});


test/lib_commands_url_TEST.js
'use strict';

var expect = require('chai').expect;
var Url = require('../lib/commands/url.js');

describe('dalek-driver-sauce Commands Url', function() {

  it('i should think of proper testing', function(){
    expect(Url).to.be.ok;
  });

});


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  // check task runtime
  require('time-grunt')(grunt);

  // load generic configs
  var configs = require('dalek-build-tools');

  grunt.initConfig({

    // load module meta data
    pkg: grunt.file.readJSON('package.json'),

    // define a src set of files for other tasks
    src: {
      lint: ['Gruntfile.js', 'index.js', 'lib/**/*.js', 'test/*.js'],
      complexity: ['index.js', 'lib/**/*.js'],
      test: ['test/*.js'],
      src: ['index.js', 'lib/**/*.js']
    },

    // clean automatically generated helper files & docs
    clean: configs.clean,

    // speed up build by defining concurrent tasks
    concurrent: configs.concurrent,

    // linting
    jshint: configs.jshint,

    // testing
    mochaTest: configs.mocha,

    // code metrics
    complexity: configs.complexity,
    plato: configs.plato(grunt.file.readJSON('.jshintrc')),

    // api docs
    yuidoc: configs.yuidocs(),

    // up version, tag & commit
    bump: configs.bump({
      pushTo: 'git@github.com:dalekjs/dalek-driver-sauce.git',
      files: ['package.json', 'CONTRIBUTORS.md', 'CHANGELOG.md']
    }),

    // generate contributors file
    contributors: configs.contributors,

    // compress artifacts
    compress: configs.compress,

    // user docs
    documantix: {
      options: {
        header: 'dalekjs/dalekjs.com/master/assets/header.html',
        footer: 'dalekjs/dalekjs.com/master/assets/footer.html',
        target: 'report/docs',
        vars: {
          title: 'DalekJS - Documentation - Driver Sauce Labs',
          desc: 'DalekJS - Documentation - Driver Sauce Labs',
          docs: true
        }
      },
      src: ['index.js']
    },

    // prepare files for grunt-plato to
    // avoid error messages (weird issue...)
    preparePlato: {
      options: {
        folders: [
          'coverage',
          'report',
          'report/complexity',
          'report/complexity/files',
          'report/complexity/files/test',
          'report/complexity/files/index_js',
          'report/complexity/files/lib_browser_js',
          'report/complexity/files/lib_commands_cookie_js',
          'report/complexity/files/lib_commands_element_js',
          'report/complexity/files/lib_commands_frame_js',
          'report/complexity/files/lib_commands_page_js',
          'report/complexity/files/lib_commands_screenshot_js',
          'report/complexity/files/lib_commands_window_js',
          'report/complexity/files/lib_commands_url_js'
        ],
        files: [
          'report.history.json',
          'files/report.history.json',
          'files/index_js/report.history.json',
          'files/lib_commands_cookie_js/report.history.json',
          'files/lib_browser_js/report.history.json',
          'files/lib_commands_element_js/report.history.json',
          'files/lib_commands_frame_js/report.history.json',
          'files/lib_commands_page_js/report.history.json',
          'files/lib_commands_screenshot_js/report.history.json',
          'files/lib_commands_url_js/report.history.json',
          'files/lib_commands_window_js/report.history.json'
        ]
      }
    },

    // prepare files & folders for coverage
    prepareCoverage: {
      options: {
        folders: ['coverage', 'report', 'report/coverage'],
        pattern: '[require("fs").realpathSync(__dirname + "/../index.js"), require("fs").realpathSync(__dirname + "/../lib/")]'
      }
    },

    // add current timestamp to the html document
    includereplace: {
      dist: {
        options: {
          globals: {
            timestamp: '<%= grunt.template.today("dddd, mmmm dS, yyyy, h:MM:ss TT") %>'
          },
        },
        src: 'report/docs/*.html',
        dest: './'
      }
    },

    // archive docs
    archive: {
      options: {
        files: ['sauce.html']
      }
    },

    // release canary version
    'release-canary': {
      options: {
        files: []
      }
    }

  });

  // load 3rd party tasks
  require('load-grunt-tasks')(grunt);
  grunt.loadTasks('./node_modules/dalek-build-tools/tasks');
  grunt.loadNpmTasks('grunt-documantix');

  // define runner tasks
  grunt.registerTask('lint', 'jshint');
  
  // split test & docs for speed
  grunt.registerTask('test', ['clean:coverage', 'prepareCoverage', 'concurrent:test', 'generateCoverageBadge']);
  grunt.registerTask('docs', ['clean:reportZip', 'clean:report', 'preparePlato', 'concurrent:docs', 'documantix', 'includereplace', 'compress']);
  
  // release tasks
  grunt.registerTask('releasePatch', ['test', 'bump-only:patch', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMinor', ['test', 'bump-only:minor', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMajor', ['test', 'bump-only:major', 'contributors', 'changelog', 'bump-commit']);
  
  // clean, test, generate docs (the CI task)
  grunt.registerTask('all', ['clean', 'test', 'docs']);
};


lib/browser.js
/*!
 *
 * Copyright (c) 2013 Sebastian Golasch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

'use strict';

// ext. libs
var Q = require('q');

/**
 * Loads the webdriver client,
 * launches the browser,
 * initializes al object properties,
 * binds to browser events
 *
 * @param {object} opts Options needed to kick off the driver
 * @constructor
 */

var DummyBrowser = function (driver) {
  this.driver = driver;
};

/**
 * Verfies a given browser config & mimics a
 * real browser while using the Sauce infrstructure
 *
 * @module Driver
 * @class Browser
 * @namespace Dalek
 */

DummyBrowser.prototype = {

  /**
   * Sauce labs authentication realm (user:key)
   *
   * @property auth
   * @type {string|null}
   */

  auth: null,

  /**
   * Browser session config
   *
   * @property sessionConfig
   * @type {object}
   */

  sessionConfig: {},

  /**
   * Saucelabs remote webdriver path
   *
   * @property path
   * @type {string}
   * @default /wd/hub
   */

  path: '/wd/hub',

  /**
   * Saucelabs remote port
   *
   * @property port
   * @type {integer}
   * @default 80
   */

  port: 80,

  /**
   * Saucelabs remote host
   *
   * @property host
   * @type {string}
   * @default host
   */

  host: 'ondemand.saucelabs.com',

  /**
   * Available browsers with their default capabilities
   *
   * @property browsers
   * @type {object}
   */

  browsers: {
    'iphone': {platform: 'OS X 10.8', version: '6', longName: 'iOS Safari (iPhone)'},
    'ipad': {platform: 'OS X 10.8', version: '6', longName: 'iOS Safari (iPad)'},
    'android': {platform: 'Linux', version: '4.0', longName: 'Chrome on Android'},
    'firefox': {platform: 'OS X 10.6', version: '25', longName: 'Mozilla Firefox'},
    'internet explorer': {platform: 'Windows 8', version: '10', longName: 'Internet Explorer'},
    'chrome': {platform: 'OS X 10.6', version: '28', longName: 'Google Chrome'},
    'opera': {platform: 'Windows 7', version: '12', longName: 'Opera'},
    'safari': {platform: 'OS X 10.8', version: '6', longName: 'Safari'}
  },

  /**
   * Default verbose browser name
   *
   * @property longName
   * @type {string}
   */

  longName: 'Mozilla Firefox',

  /**
   * Default desired capabilities
   *
   * @property desiredCapabilities
   * @type {object}
   */

  desiredCapabilities: {
    platform: 'OS X 10.8',
    browserName: 'chrome',
    'browser-version': 'latest',
    name: 'Dalek Testsuite'
  },

  /**
   * Driver defaults
   *
   * @property driverDefaults
   * @type {object}
   */

  driverDefaults: {
    viewport: false,
    status: {
      os: {
        name: 'Linux',
        version: null,
        arch: null
      }
    },
    sessionInfo: true
  },

  /**
   * Available platforms
   *
   * @property platforms
   * @type {array}
   */

  platforms: ['Windows 8.1', 'Windows 8', 'Windows 7', 'Windows XP', 'OS X 10.9', 'OS X 10.8', 'OS X 10.6', 'Linux'],

  /**
   * Stores & validates the incoming browser config
   *
   * @method launch
   * @param {object} configuration Browser configuration
   * @param {EventEmitter2} events EventEmitter (Reporter Emitter instance)
   * @param {Dalek.Internal.Config} config Dalek configuration class
   * @return {object} Browser promise
   */

  launch: function (configuration, events, config) {
    var deferred = Q.defer();

    // override desired capabilities, status & browser longname
    this.desiredCapabilities = this._generateDesiredCaps(configuration.name, config);
    this.driverDefaults.status = this._generateStatusInfo(this.desiredCapabilities);
    this.longName = this._generateLongName(configuration.name,config);

    // store injected configuration/log event handlers
    this.reporterEvents = events;
    this.configuration = configuration;
    this.config = config;

    // immediatly resolve the deferred
    deferred.resolve();
    return deferred.promise;
  },

  /**
   * Kills the remote browser
   * TODO: Close the remote session
   *
   * @method kill
   * @return {object} Promise
   */

  kill: function () {
    var deferred = Q.defer();
    deferred.resolve();
    return deferred.promise;
  },

  /**
   * Generates the status defaults for the OS configuration
   *
   * @method getStatusDefaults
   * @param {object} desiredCapabilities Desired capabilities
   * @param {object} desiredCapabilities Desired capabilities
   * @return {object} OS status information
   */

  getDesiredCapabilities: function (browserName, config) {
    return this._generateDesiredCaps(browserName, config);
  },

  /**
   * Generates the status defaults for the OS configuration
   *
   * @method getStatusDefaults
   * @param {object} desiredCapabilities Desired capabilities
   * @return {object} OS status information
   */

  getStatusDefaults: function (desiredCapabilities) {
    return this._generateStatusInfo(desiredCapabilities);
  },

  /**
   * Sets the sauce authentication token
   *
   * @method setAuth
   * @param {string} user Sauce labs username
   * @param {string} key Sauce labs key
   * @chainable
   */

  setAuth: function (user, key) {
    this.auth = user + ':' + key;
    return this;
  },

  /**
   * Verifies the browser config
   *
   * @method _verfiyBrowserConfig
   * @param {string} browserName Name of the browser to verify
   * @param {object} config Daleks internal config helper
   * @return {object} Browser config
   * @private
   */

  _verfiyBrowserConfig: function (browserName, config) {
    var browsers = config.get('browsers');

    // check if we couldnt find a configured alias,
    // set to defaults otherwise
    if (!browsers) {
      return {actAs: this.desiredCapabilities.browserName, version: this.desiredCapabilities['browser-version']};
    }
    var browser = config.get('browsers')[0][browserName] || null;
    // check if we couldnt find a configured alias,
    // check and apply if there is a default config
    if (!browser && this.browsers[browserName]) {
      browser = this.browsers[browserName];
    }

    // check if the actas property has been set, if not
    // use the given browser name
    if (!browser.actAs) {
      browser.actAs = browserName;
    }

    return browser;
  },

  /**
   * Verfies the OS platform config
   *
   * @method _verfiyPlatformConfig
   * @param {object} browser Browser information
   * @return {string} Platform
   * @private
   */

  _verfiyPlatformConfig: function (browser) {
    var isValid = this.platforms.reduce(function (previousValue, platform) {
      if (previousValue === browser.platform || platform === browser.platform) {
        return browser.platform;
      }
    });

    return isValid || this.desiredCapabilities.platform;
  },

  /**
   * Generates the desired capabilities for this session
   *
   * @method _generateDesiredCaps
   * @param {string} browserName The browser name
   * @param {object} config Daleks internal config helper
   * @return {object} The sessions desired capabilities
   * @private
   */

  _generateDesiredCaps: function (browserName, config) {
    var browser = this._verfiyBrowserConfig(browserName, config);
    var platform = this._verfiyPlatformConfig(browser);
    var driverConfig = config.get('driver.sauce');
    var desiredCaps = {
      browserName: browser.actAs,
      platform: platform,
      'screen-resolution': (browser['screen-resolution'] || null),
      version: (browser.version || this.desiredCapabilities['browser-version']),
      name: driverConfig.name || this.desiredCapabilities.name
    };

    // check if the user added tags
    if (driverConfig.tags) {
      desiredCaps.tags = driverConfig.tags;
    }

    // check if the user added a build id
    if (driverConfig.build) {
      desiredCaps.build = driverConfig.build;
    }

    return desiredCaps;
  },

  /**
   * Generates OS status information
   *
   * @method _generateStatusInfo
   * @param {object} desiredCaps The sessions desired capabilities
   * @return {object} OS status information
   * @private
   */

  _generateStatusInfo: function (desiredCaps) {
    return {os: {name: desiredCaps.platform, arch: null, version: null}};
  },

  /**
   * Generates the verbose name of the current remote browser in use
   *
   * @method _generateLongName
   * @param {object} desiredCaps The sessions desired capabilities
   * @return {string} Verbose browser name
   * @private
   */

  _generateLongName: function (browserName, config) {
    var longName = null;
    if(this.browsers.hasOwnProperty(browserName)){
      longName = this.browsers[browserName].longName;
    }
    if(config.get('browsers')[0].hasOwnProperty(browserName)){
      longName = config.get('browsers')[0][browserName].longName;
    }
    return longName;
  }
};

module.exports = DummyBrowser;


index.js
/*!
 *
 * Copyright (c) 2013 Sebastian Golasch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

'use strict';

// ext. libs
var fs = require('fs');
var Q = require('q');

// int. libs
var WD = require('dalek-internal-webdriver');
var Browser = require('./lib/browser');

/**
 * Initializes the sauce labs driver & the remote browser instance
 *
 * @param {object} opts Initializer options
 * @constructor
 */

var Sauce = function (opts) {
  // get the browser configuration & the browser module
  var browserConf = {name: null};
  var browser = opts.browserMo;

  // prepare properties
  this._initializeProperties(opts);

  // create a new webdriver client instance
  this.webdriverClient = new WD(browser, this.events);

  // listen on browser events
  this._startBrowserEventListeners(browser);

  // assign the current browser name
  browserConf.name = this.browserName;

  // store desired capabilities of this session
  this.desiredCapabilities = browser.getDesiredCapabilities(this.browserName, this.config);
  this.browserDefaults = browser.driverDefaults;
  this.browserDefaults.status = browser.getStatusDefaults(this.desiredCapabilities);
  this.browserData = browser;

  // set auth data
  var driverConfig = this.config.get('driver.sauce');
  browser.setAuth(driverConfig.user, driverConfig.key);

  // launch the browser & when the browser launch
  // promise is fullfilled, issue the driver:ready event
  // for the particular browser
  browser
    .launch(browserConf, this.reporterEvents, this.config)
    .then(this.events.emit.bind(this.events, 'driver:ready:sauce:' + this.browserName, browser));
};

/**
 * This module is a driver plugin for [DalekJS](//github.com/dalekjs/dalek).
 * It connects Daleks testsuite with the remote testing environment of [Sauce Labs](https://saucelabs.com).
 *
 * The driver can be installed with the following command:
 *
 * ```bash
 * $ npm install dalek-driver-sauce --save-dev
 * ```
 *
 * You can use the driver by adding a config option to the your [Dalekfile](/docs/config.html)
 *
 * ```javascript
 * "driver": ["sauce"]
 * ```
 *
 * Or you can tell Dalek that it should run your tests via sauces service via the command line:
 *
 * ```bash
 * $ dalek mytest.js -d sauce
 * ```
 *
 * In order to run your tests within the Sauce Labs infrastructure, you must add your sauce username & key 
 * to your dalek configuration. Those two parameters must be set in order to get this driver up & running.
 * You can specifiy them within your [Dalekfile](/docs/config.html) like so:
 *
 * ```javascript
 * "driver.sauce": {
 *   "user": "dalekjs",
 *   "key": "aaaaaa-1234-567a-1abc-1br6d9f68689"
 * }
 * ```
 *
 * It is also possible to specify a set of other extra saucy parameters like `name` & `tags`:
 *
 * ```javascript
 * "driver.sauce": {
 *   "user": "dalekjs",
 *   "key": "aaaaaa-1234-567a-1abc-1br6d9f68689",
 *   "name": "Guineapig",
 *   "tags": ["dalek", "testproject"]
 * }
 * ```
 *
 * If you would like to have a more control over the browser/OS combinations that are available, you are able 
 * to configure you custom combinations:
 *
 * ```javascript
 * "browsers": [{
 *   "chrome": {
 *     "platform": "OS X 10.6",
 *     "actAs": "chrome",
 *     "version": 27
 *   },
 *   "chromeWin": {
 *     "platform": "Windows 7",
 *     "actAs": "chrome",
 *     "version": 27
 *   },
 *   "chromeLinux": {
 *     "platform": "Linux",
 *     "actAs": "chrome",
 *     "version": 26
 *   }
 * ```
 * 
 * You can then call your custom browsers like so:
 *
 * ```bash
 * $ dalek mytest.js -d sauce -b chrome,chromeWin,chromeLinux
 * ```
 *
 * or you can define them in your Dalekfile:
 *
 * ```javascript
 * "browser": ["chrome", "chromeWin", "chromeLinux"]
 * ```
 *
 * A list of all available browser/OS combinations, can be found [here](https://saucelabs.com/docs/platforms).
 * 
 * @module Driver
 * @class Sauce
 * @namespace Dalek
 * @part Sauce
 * @api
 */

Sauce.prototype = {

  /**
   * Initializes the driver properties
   *
   * @method _initializeProperties
   * @param {object} opts Options needed to kick off the driver
   * @chainable
   * @private
   */

  _initializeProperties: function (opts) {
    // prepare properties
    this.actionQueue = [];
    this.config = opts.config;
    this.lastCalledUrl = null;
    this.driverStatus = {};
    this.sessionStatus = {};
    // store injcted options in object properties
    this.events = opts.events;
    this.reporterEvents = opts.reporter;
    this.browserName = opts.browser;
    return this;
  },

  /**
   * Binds listeners on browser events
   *
   * @method _initializeProperties
   * @param {object} browser Browser module
   * @chainable
   * @private
   */

  _startBrowserEventListeners: function (browser) {
    // issue the kill command to the browser, when all tests are completed
    this.events.on('tests:complete:sauce:' + this.browserName, browser.kill.bind(browser));
    // clear the webdriver session, when all tests are completed
    this.events.on('tests:complete:sauce:' + this.browserName, this.webdriverClient.closeSession.bind(this.webdriverClient));
    return this;
  },

  /**
   * Checks if a webdriver session has already been established,
   * if not, create a new one
   *
   * @method start
   * @return {object} promise Driver promise
   */

  start: function () {
    var deferred = Q.defer();

    // store desired capabilities of this session
    this.desiredCapabilities = this.browserData.getDesiredCapabilities(this.browserName, this.config);
    this.browserDefaults = this.browserData.driverDefaults;
    this.browserDefaults.status = this.browserData.getStatusDefaults(this.desiredCapabilities);

    // start a browser session
    this._startBrowserSession(deferred, this.desiredCapabilities, this.browserDefaults);

    return deferred.promise;
  },

  /**
   * Creates a new webdriver session
   * Gets the driver status
   * Gets the session status
   * Resolves the promise (e.g. let them tests run)
   *
   * @method _startBrowserSession
   * @param {object} deferred Browser session deferred
   * @chainable
   * @private
   */

  _startBrowserSession: function (deferred, desiredCapabilities, defaults) {
    var viewport = this.config.get('viewport');

    // start a session, transmit the desired capabilities
    var promise = this.webdriverClient.createSession({desiredCapabilities: desiredCapabilities});
    
    // set the default viewport if supported by the browser
    if (defaults.viewport) {
      promise = promise.then(this.webdriverClient.setWindowSize.bind(this.webdriverClient, viewport.width, viewport.height));
    }

    // get the driver status if supported by the browser
    if (defaults.status === true) {
      promise = promise
        .then(this.webdriverClient.status.bind(this.webdriverClient))
        .then(this._driverStatus.bind(this));
    } else {
      promise = promise.then(this._driverStatus.bind(this, JSON.stringify({value: defaults.status})));
    }

    // get the session info if supported by the browser
    if (defaults.sessionInfo === true) {
      promise = promise
        .then(this.webdriverClient.sessionInfo.bind(this.webdriverClient))
        .then(this._sessionStatus.bind(this));
    } else {
      promise = promise.then(this._driverStatus.bind(this, JSON.stringify({value: defaults.sessionInfo})));
    }

    // finally resolve the deferred
    promise.then(deferred.resolve);
    return this;
  },

  /**
   * Starts to execution of a batch of tests
   *
   * @method end
   * @chainable
   */

  end: function () {
    var result = Q.resolve();

    // loop through all promises created by the remote methods
    // this is synchronous, so it waits if a method is finished before
    // the next one will be executed
    this.actionQueue.forEach(function (f) {
      result = result.then(f);
    });

    // flush the queue & fire an event
    // when the queue finished its executions
    result.then(this.flushQueue.bind(this));
    return this;
  },

  /**
   * Flushes the action queue (e.g. commands that should be send to the wbdriver server)
   *
   * @method flushQueue
   * @chainable
   */

  flushQueue: function () {
    // clear the action queue
    this.actionQueue = [];
    
    // kill the session
    var promise = this.webdriverClient.deleteSession();

    // emit the run.complete event
    promise.then(function () {
      this.events.emit('driver:message', {key: 'run.complete', value: null});
    }.bind(this));

    return this;
  },

  /**
   * Loads the browser session status
   *
   * @method _sessionStatus
   * @param {object} sessionInfo Session information
   * @return {object} promise Browser session promise
   * @private
   */

  _sessionStatus: function (sessionInfo) {
    var defer = Q.defer();
    this.sessionStatus = JSON.parse(sessionInfo).value;
    this.events.emit('driver:sessionStatus:sauce:' + this.browserName, this.sessionStatus);
    defer.resolve();
    return defer.promise;
  },

  /**
   * Loads the browser driver status
   *
   * @method _driverStatus
   * @param {object} statusInfo Driver status information
   * @return {object} promise Driver status promise
   * @private
   */

  _driverStatus: function (statusInfo) {
    var defer = Q.defer();
    this.driverStatus = JSON.parse(statusInfo).value;
    this.events.emit('driver:status:sauce:' + this.browserName, this.driverStatus);
    defer.resolve();
    return defer.promise;
  },

  /**
   * Creates an anonymus function that calls a webdriver
   * method that has no return value, emits an empty result event
   * if the function has been run
   * TODO: Name is weird, should be saner
   *
   * @method _createNonReturnee
   * @param {string} fnName Name of the webdriver function that should be called
   * @return {function} fn
   * @private
   */

  _createNonReturnee: function (fnName) {
    return this._actionQueueNonReturneeTemplate.bind(this, fnName);
  },

  /**
   * Generates a chain of webdriver calls for webdriver
   * methods that don't have a return value
   * TODO: Name is weird, should be saner
   *
   * @method _actionQueueNonReturneeTemplate
   * @param {string} fnName Name of the webdriver function that should be called
   * @param {string} hash Unique action hash
   * @param {string} uuid Unique action hash
   * @chainable
   * @private
   */

  _actionQueueNonReturneeTemplate:function (fnName, hash, uuid) {
    this.actionQueue.push(this.webdriverClient[fnName].bind(this.webdriverClient));
    this.actionQueue.push(this._generateDummyDriverMessageFn.bind(this, fnName, hash, uuid));
    return this;
  },

  /**
   * Creates a driver notification with an empty value
   * TODO: Name is weird, should be saner
   *
   * @method _generateDummyDriverMessageFn
   * @param {string} fnName Name of the webdriver function that should be called
   * @param {string} hash Unique action hash
   * @param {string} uuid Unique action hash
   * @return {object} promise Driver message promise
   * @private
   */

  _generateDummyDriverMessageFn: function (fnName, hash, uuid) {
    var deferred = Q.defer();
    this.events.emit('driver:message', {key: fnName, value: null, uuid: uuid, hash: hash});
    deferred.resolve();
    return deferred.promise;
  }
};

/**
 * Determines if the driver is a "multi" browser driver,
 * e.g. can handle more than one browser
 *
 * @method isMultiBrowser
 * @return {bool} isMultiBrowser Driver can handle more than one browser
 */

module.exports.isMultiBrowser = function () {
  return true;
};

/**
 * Verifies a browser request
 * TODO: Still a noop, need to add "verify the browser" logic
 *
 * @method verifyBrowser
 * @return {bool} isVerifiedBrowser Driver can handle this browser
 */

module.exports.verifyBrowser = function () {
  return true;
};

/**
 * Determines if the driver comes with its own browsers bundled
 *
 * @method dummyBrowser
 * @return {bool} isDummyBrowser Driver does not rely on Dalek browser modules
 */

module.exports.dummyBrowser = function () {
  return true;
};

/**
 * Returnes the browser that comes bundled with the driver
 *
 * @method getBrowser
 * @param {object} driver Driver instance
 * @return {object} Browser module
 */

module.exports.getBrowser = function (driver) {
  return {configuration: null, module: new Browser(driver)};
};

/**
 * Creates a new driver instance
 *
 * @method create
 * @param {object} opts Options needed to kick off the driver
 * @return {Sauce} driver
 */

module.exports.create = function (opts) {
  // load the remote command helper methods
  var dir = __dirname + '/lib/commands/';
  fs.readdirSync(dir).forEach(function (file) {
    require(dir + file)(Sauce);
  });

  return new Sauce(opts);
};
